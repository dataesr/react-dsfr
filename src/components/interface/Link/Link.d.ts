import * as React from 'react';

export type LinkClassName = string | Object | any[];

export type LinkChildren = React.ReactNode[] | React.ReactNode | string;

export type LinkVerticalIconPosition = "top" | "middle";

export type LinkIconPosition = "left" | "right";

export type LinkIconSize = "fw" | "xxs" | "xs" | "sm" | "1x" | "lg" | "lg" | "xl" | "2x" | "3x" | "10x";

export type LinkDisplay = "inline" | "flex";

export interface LinkProps {
    className?: LinkClassName;
    children?: LinkChildren;
    href?: string;
    verticalIconPosition?: LinkVerticalIconPosition;
    as?: React.ReactElement<any>;
    title?: string;
    target?: string;
    isSimple?: boolean;
    current?: boolean;
    icon?: string;
    onClick?: (...args: any[])=>any;
    iconPosition?: LinkIconPosition;
    iconSize?: LinkIconSize;
    /**
     * @ignore
     */
    display?: LinkDisplay;
}

declare const Link: React.FC<LinkProps>;

export default Link;

