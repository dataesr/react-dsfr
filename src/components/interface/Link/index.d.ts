import type Link from './Link';

export type { Link };
