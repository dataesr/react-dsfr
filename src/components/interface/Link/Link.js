import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Icon from '../../foundation/icon/index';
import dataAttributes from '../../../utils/data-attributes';
import useRouterContext from '../../../hooks/useRouterContext';

import '@gouvfr/dsfr/dist/css/links.min.css';

/**
 *
 * @visibleName Link
 */
const Link = ({
  children,
  href,
  to,
  title,
  target,
  isSimple,
  className,
  as,
  current,
  display,
  iconSize,
  icon,
  iconPosition,
  onClick,
  ...remainingProps
}) => {
  const Router = useRouterContext();
  const _className = classNames(
    className, {
      [`ds-fr--${display}`]: display && icon,
      'fr-link': isSimple,
    },
  );
  const Component = (Router && to) ? Router : 'a';
  const onClickLink = (e) => {
    e.preventDefault();
    onClick();
  };

  const _link = (
    <Component
      onClick={onClick ? (e) => onClickLink(e) : undefined}
      aria-current={current ? 'page' : undefined}
      href={href || undefined}
      to={to || undefined}
      title={title || undefined}
      target={target}
      rel={(target === '_blank') ? 'noopener noreferrer' : undefined}
      className={_className}
      {...dataAttributes(remainingProps)}
    >
      {children}
    </Component>
  );
  return icon ? (
    <Icon
      className={classNames({ 'ds-fr--v-top': display && icon })}
      name={icon}
      size={iconSize}
      iconPosition={_link.props?.children ? iconPosition : 'center'}
    >
      {_link}
    </Icon>
  ) : _link;
};

Link.defaultProps = {
  className: '',
  title: '',
  target: '_self',
  isSimple: false,
  current: false,
  icon: '',
  as: null,
  iconPosition: 'right',
  href: '',
  to: '',
  children: '',
  onClick: null,
  display: 'inline',
  iconSize: '1x',
};

Link.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.string,
  ]),
  href: PropTypes.string,
  to: PropTypes.string,
  as: PropTypes.element,
  title: PropTypes.string,
  target: PropTypes.string,
  isSimple: PropTypes.bool,
  current: PropTypes.bool,
  icon: PropTypes.string,
  onClick: PropTypes.func,
  iconPosition: PropTypes.oneOf(['left', 'right']),
  iconSize: PropTypes.oneOf(['fw', 'xxs', 'xs', 'sm', '1x', 'lg', 'lg', 'xl', '2x', '3x', '10x']),
  /**
   * @ignore
   */
  display: PropTypes.oneOf(['inline', 'flex']),
};

export default Link;
