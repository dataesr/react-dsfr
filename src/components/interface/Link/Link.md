## Lien - Link

Le lien permet la navigation. Pour les actions d’un autre type, comme la soumission d’un formulaire, il faut utiliser le composant boutons.

### Lien contextuel
Le lien contextuel se trouve au sein d’un texte et reprend les caractéristiques typographiques de celui-ci (font, couleur, taille) tout en étant souligné.

```jsx
<Link
    title="titre lien interne"
    href="/">lien contextuel
</Link>
```

```jsx
<Link
    icon="ri-arrow-go-back-line"
    iconPosition="left"
    title="titre lien interne" href="/my-page">
    lien contextuel avec icône
</Link>
```

```jsx
<Link
    icon="ri-arrow-go-back-line"
    title="titre lien interne"
    href="/my-page">
    lien contextuel avec icône
</Link>
```

#### Usage avec react-router-dom

```jsx
import { BrowserRouter as Router, Route, Link as RouterLink } from 'react-router-dom';
import { RouterContextProvider} from '../../../hooks/useRouterContext';
// Wrap the entire app with RouterContextProvider
<RouterContextProvider Router={RouterLink}>
<Router>
    <Route
        path="/"
        component={(props) => (
            <>
              <Link href="https://dataesr.fr">external</Link>
               //
              <Link to="/">internal</Link>
            </>
        )}
    />
</Router>
</RouterContextProvider>
```

### Lien externe
Le lien ‘externe’ doit être utilisé pour les liens pointant vers un autre site et ouvrant un nouvel onglet (via l’attribut target=”blank”).

```jsx
<Link
    title="titre lien externe - nouvelle fenêtre"
    href="/"
    target="_blank">lien externe - nouvelle fenêtre
</Link>
```

#### Lien simple

```jsx
<Link
    href="/my-page"
    isSimple>lien simple
</Link>
```

#### Usage avec react-router-dom

```jsx
import { BrowserRouter as Router, Route, Link as RouterLink } from 'react-router-dom';

<Router>
    <Route
        path="/"
        component={(props) => (
            <Link
                isSimple
                as={<RouterLink to="/my-page">lien simple</RouterLink>}
            />
        )}
    />
</Router>
```

```jsx
<Link
    href="/"
    isSimple
    icon="ri-arrow-right-line">lien simple avec icône à droite
</Link>
```

```jsx
<Link
    href="/"
    isSimple
    icon="ri-arrow-left-line"
    iconPosition="left">lien simple avec icône à gauche
</Link>
```

L’icône seule n'est à utiliser que très rarement, et le lien doit contenir un texte afin d'être lu par les lecteurs d'écran.
```jsx
<Link
    href="/"
    isSimple
    icon="ri-arrow-right-up-line"
/>
```
