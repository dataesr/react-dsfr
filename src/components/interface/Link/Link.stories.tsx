import { Story } from '@storybook/react/types-6-0';

import Link, { LinkProps } from './Link';

// https://www.systeme-de-design.gouv.fr/elements-d-interface/composants/lien

export default {
  title: 'Composants/Lien - Link',
  component: Link,
  parameters: {
    docs: {
      description: {
        component: 'Le lien permet la navigation entre une page et un autre contenu au sein de la même page, du même site ou externe.',
      },
    },
  },
};

const Template: Story<LinkProps> = (args) => <Link {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Votre lien',
  href: 'https://www.example.com',
};
Default.storyName = 'Lien';

export const LinkInText = Template.bind({});
LinkInText.args = {
  children: (
    <p>
      {' '}
      Votre
      {' '}
      <Link href="https://www.example.com">lien</Link>
    </p>
  ),
};
LinkInText.storyName = 'Lien au fil du texte';

export const Target = Template.bind({});
Target.args = {
  children: "Lien vers l'exterieur",
  href: 'https://www.example.com',
  target: '_blank',
};
Target.storyName = "Lien vers l'exterieur";

export const Size = Template.bind({});
Size.args = {
  children: 'Petit',
  href: 'https://www.example.com',
  size: 'sm',
};
Size.storyName = 'Taille du lien';

export const Title = Template.bind({});
Title.args = {
  children: 'Votre lien avec description',
  href: 'https://www.example.com',
  title: 'Vers le lien https://www.example.com',
};
Title.storyName = 'Description du lien';

export const LeftIcon = Template.bind({});
LeftIcon.args = {
  children: 'Lien avec icône à gauche',
  href: 'https://www.example.com',
  icon: 'fr-fi-arrow-right-line',
  iconPosition: 'left',
};
LeftIcon.storyName = 'Lien avec icône à gauche';

export const RightIcon = Template.bind({});
RightIcon.args = {
  children: 'Lien avec icône à droite',
  href: 'https://www.example.com',
  icon: 'fr-fi-arrow-right-line',
  iconPosition: 'right',
};
RightIcon.storyName = 'Lien avec icône à droite';

export const IconSize = Template.bind({});
IconSize.args = {
  children: 'Grande icône sur lien',
  href: 'https://www.example.com',
  icon: 'fr-fi-arrow-right-line',
  iconSize: 'xl',
  iconPosition: 'right',
};
IconSize.storyName = "Taille de l'icône";

export const IconVerticalPosition = Template.bind({});
IconVerticalPosition.args = {
  children: 'Icône en position vers le haut',
  href: 'https://www.example.com',
  icon: 'fr-fi-arrow-right-line',
  verticalIconPosition: 'top',
};
IconVerticalPosition.storyName = "Position vertical de l'icône";
