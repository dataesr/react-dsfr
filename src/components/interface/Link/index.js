import '@gouvfr/dsfr/dist/css/links.min.css';

export { default } from './Link';
