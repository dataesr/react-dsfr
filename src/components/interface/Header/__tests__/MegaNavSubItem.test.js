import renderer from 'react-test-renderer';
import { MegaNavSubItem } from '../index';
import Link from '../../Link';

describe('<MegaNavSubItem />', () => {
  const initialProps = {
    title: 'title',
    link: 'link',
  };

  it('renders correctly', () => {
    const component = renderer
      .create(
        <MegaNavSubItem title={initialProps.title} link={initialProps.link}>
          <Link title="title" href="/">Link #1</Link>
        </MegaNavSubItem>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
