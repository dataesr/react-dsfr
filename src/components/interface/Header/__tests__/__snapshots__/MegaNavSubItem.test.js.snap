// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MegaNavSubItem /> renders correctly 1`] = `
<div
  className="fr-col-12 fr-col-lg-3"
>
  <h5
    className="fr-mega-menu__category"
  >
    <a
      className="fr-nav__link"
      target="_self"
    >
      title
    </a>
  </h5>
  <ul
    className="fr-mega-menu__list"
  >
    <li>
      <a
        className="fr-nav__link"
        href="/"
        target="_self"
        title="title"
      >
        Link #1
      </a>
    </li>
  </ul>
</div>
`;
